NIFI-6009 ScanKudu Processor & KuduPut Processor Delete Operation

SandishKumarHN opened the pull request:
JIRA: NIFI-6009
- ScanKudu Processor
- PutKudu Delete Operation
- Kudu Client Service

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [x] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [x] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [x] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

SandishKumarHN commented:
@cammachusa I see that you worked PutKudu Processor, I have some changes to PutKudu and also Added ScanKudu Processor. please review and let me know if I have missed anything.

SandishKumarHN commented:
@joewitt @bbende can you please have a look into this.

bbende commented:
Sorry but I am not familiar with Kudu at all

SandishKumarHN commented:
> Sorry but I am not familiar with Kudu at all

Okay np, please give me feedback on NIFI part.

granthenke commented on nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-processors/src/main/java/org/apache/nifi/kudu/PutKudu.java:
```diff
@@ -279,43 +217,56 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
 
-    private void trigger(final ProcessContext context, final ProcessSession session, final List<FlowFile> flowFiles) throws ProcessException {
-        final KuduSession kuduSession = getKuduSession(kuduClient);
+    private void trigger(final ProcessContext context, final ProcessSession session, final List<FlowFile> flowFiles,  KuduClientService clientService) throws ProcessException {
```
Can the clientService be a class variable so it doesn't need to be passed? Similar to the `kuduClient` before clientService was added.

granthenke commented on nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-processors/src/main/java/org/apache/nifi/kudu/PutKudu.java:
```diff
@@ -131,8 +115,9 @@
     protected static final PropertyDescriptor INSERT_OPERATION = new Builder()
```
I am not sure the best way for Nifi to handle a change like this, but calling this `INSERT_OPERATION` while also supporting DELETE doesn't make a lot of sense. It should probably be something like `OPERATION_TYPE`.

granthenke commented on nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-client-service-api/src/main/java/org/apache/nifi/kudu/OperationType.java:
```diff
@@ -15,10 +15,11 @@
     INSERT,
     INSERT_IGNORE,
-    UPSERT;
+    UPSERT,
```
What about `UPDATE` operations?

granthenke commented on nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-processors/src/main/java/org/apache/nifi/kudu/PutKudu.java:
```diff
@@ -279,43 +217,56 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
         int numBuffered = 0;
         final List<RowError> pendingRowErrors = new ArrayList<>();
         for (FlowFile flowFile : flowFiles) {
+            operationType = OperationType.valueOf(context.getProperty(INSERT_OPERATION).evaluateAttributeExpressions(flowFile).getValue());
```
Is the idea that a user would have a flow of all INSERTS or all DELETE operations? Are there example of other nifi processors that handle different operation types? (I am not super familiar with Nifi)

SandishKumarHN replied:
Yes, the idea is that a user would have a flow of all INSERTS or all DELETE operations

granthenke commented on nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-processors/src/main/java/org/apache/nifi/kudu/PutKudu.java:
```diff
@@ -279,43 +217,56 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
-                    Operation operation = operationType == OperationType.UPSERT
-                        ? upsertRecordToKudu(kuduTable, record, fieldNames)
-                        : insertRecordToKudu(kuduTable, record, fieldNames);
+                    Operation operation = operationType == OperationType.DELETE
```
This is a pretty complicated ternary statement, it would probably be easier to read as a switch statement or an if/else.

granthenke commented on nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-processors/src/main/java/org/apache/nifi/kudu/PutKudu.java:
```diff
@@ -403,92 +354,21 @@ protected KuduSession getKuduSession(final KuduClient client) {
-        } catch (Exception ex) {
-            return -1;
-        }
+    protected Delete deleteRecordToKudu(KuduClientService clientService, KuduTable kuduTable, Record record, List<String> fieldNames) {
```
It's worth noting that until Kudu 1.10 releases, delete operations can only have the key columns set, if any additional column is set the operation will be failed. 

This is the relevant commit: 
https://github.com/apache/kudu/commit/17fb5a10319d2bc19e8e54aa9c088335b04926e9#diff-7e3732d5ab4a5b01ff1955907bfc82f8

SandishKumarHN replied:
NOTED

granthenke commented on nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-client-services/src/main/java/org/apache/nifi/kudu/KuduControllerService.java:
```diff
@@ -0,0 +1,344 @@
+    public void buildPartialRow(Schema schema, PartialRow row, Record record, List<String> fieldNames) {
+        for (String colName : fieldNames) {
+            int colIdx = this.getColumnIndex(schema, colName);
+            if (colIdx != -1) {
```
What happens to missing columns?

SandishKumarHN replied:
it will be Ignored. it only takes fields which are in kudu table

granthenke commented on nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-client-services/src/main/java/org/apache/nifi/kudu/KuduControllerService.java:
```diff
@@ -0,0 +1,344 @@
+                ColumnSchema colSchema = schema.getColumnByIndex(colIdx);
+                Type colType = colSchema.getType();
+
+                if (record.getValue(colName) == null) {
```
Does Nifi separate the concept of null vs unset? 

Because Kudu offers the ability for default values we have seen other integrations get tripped up on this. See Spark for example: https://github.com/apache/kudu/blob/master/java/kudu-spark/src/main/scala/org/apache/kudu/spark/kudu/RowConverter.scala#L52

granthenke commented on nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-processors/src/main/java/org/apache/nifi/kudu/io/RowSerializer.java:
```diff
@@ -0,0 +1,44 @@
+import java.io.OutputStream;
+import java.util.Iterator;
+
+public interface RowSerializer {
```
What is this used for? Why is it needed?

granthenke commented on nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-processors/src/main/java/org/apache/nifi/kudu/io/JsonFullRowSerializer.java:
```diff
@@ -0,0 +1,119 @@
+ *    ]
+ * }
+ */
+public class JsonFullRowSerializer implements RowSerializer {
```
What is this used for? Why is it needed?

SandishKumarHN replied:
Removed

granthenke commented on nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-processors/src/main/java/org/apache/nifi/kudu/io/JsonFullRowSerializer.java:
```diff
@@ -0,0 +1,119 @@
+            while (columns.hasNext()) {
+                ColumnSchema col = columns.next();
+                jsonBuilder.append("\"" + col.getName() + "\":");
+                switch (col.getType()) {
```
Note in the next release of Kudu, if you don't care about the type returned you can use `result.getObject`.
https://github.com/apache/kudu/commit/562b5f12f721b533cc2cb868c15a2f85d9e0114b#diff-d622d57293d3d23f774030d3c5916b09

SandishKumarHN replied:
NOTED

granthenke commented on nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-client-services/src/main/java/org/apache/nifi/kudu/KuduControllerService.java:
```diff
@@ -0,0 +1,344 @@
+        }
+    }
+
+    protected void addPredicate(AsyncKuduScanner.AsyncKuduScannerBuilder scannerBuilder, KuduTable kuduTable, String column, String value) {
```
Is there a reason to only support `Equals` predicates?

SandishKumarHN replied:
For now, I kept it only Equals predicates, I will create separate JIRA for adding more predicted support, right now I'm not sure about NIFI expressions supported for other predicates.

granthenke commented on nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-client-services/src/main/java/org/apache/nifi/kudu/KuduControllerService.java:
```diff
@@ -0,0 +1,344 @@
+                    continue;
+                }
+
+                switch (colType.getDataType(colSchema.getTypeAttributes())) {
```
You should need getDataType, switching on `Type` is sufficient.

markap14 commented:
I have fairly minimal Kudu experience but am very familiar with NiFi so between Grant and myself we should have this covered pretty well. :)

I have not been thoroughly through this PR, but I noticed a significant refactoring into creating a Controller Service. While we do have some processors that use a "client" type of Controller Service, we have to be very careful modifying an existing processor like this to change how this is configured. Doing so would mean that everyone who uses the Kudu processor currently would have an invalid flow when they upgrade and would then have to completely reconfigure everything. We would need to have a really good, defensible argument for changing the processor in that way. Can you explain the rationale behind such a change?

It also appears that you changed the package name from "org.apache.nifi.processors.kudu" to "org.apache.nifi.kudu" - we cannot make a change like this, as it would result in processors that exist in current flows becoming "Ghost Processors."

I also wonder if it makes more sense to have a DeleteKudu processor. PutKudu implies to me that it is going to put data into Kudu, not delete data...

SandishKumarHN commented:
@granthenke that's a fantastic review, will update PR with suggested changes.

@markap14 

1) added Kudu Controller Service to reuse Kudu client connection's code and have a single Kudu client for multiple Kudu Processor in a pipeline. I thought Its an early stage for PutKudu Processor good time to make change 
2) Will rename back to org.apache.nifi.processors.kudu
3) we could have a separate DeleteKudu processor, reason for adding DELETE in Put processor to overcome particular race condition. i.e CDC operations, with a separate Kudu Delete Processor INSERT would go first and next coming DELETE would drop inserted records or vice-versa.

markap14 commented:
@SandishKumarHN Thanks for responding so quickly.
Re: the Controller Service - @granthenke can correct me if I am wrong, but I don't know that reusing the client provides a huge upside here. It probably doesn't hurt but I don't know that we really gain much either. In terms of code reuse, I think we can easily accomplish that in other ways, perhaps with an abstract class or helper methods. A change that makes a flow invalid is pretty much a last resort. If there were some sort of security vulnerability or something like that, or if we find that we're doing something egregious to the server side then it would make sense to consider the change.

Can you explain your third thought there, regarding CDC operations? Can you outline the use case where this helps? As you're describing it, I think such a race condition could still exist. If you wanted to INSERT some records and DELETE some records, you'd still need a PutKudu processor configured to INSERT and another PutKudu Processor configured to DELETE - unless I'm missing something?

SandishKumarHN commented:
@markap14 
1) I can certainly create an abstract or helper class to reuse of Kudu Client Connection.

for the 3) The race condition would happen when there is an order mismatch while separate  Kudu INSERT Processor, Kudu DELETE Processor, we can overcome by using a single processor with INSERT and DELETE with a flow file attribute for 'OperationType'

SandishKumarHN commented:
@granthenke @markap14 thanks for the awesome feedback. made changes according to the suggestions. Please review again.

- Renamed package back to org.apache.nifi.processors.kudu, removed RowSerializer Class and added convertToJson method instead, removed ControlService, disabled exposing AsysnKuduClient and made changes to buildPartialRow.

- Added AbstractKuduProcessor which handles kudu connections and put and scan operations

- Have added OperationType to PutKudu with Options of (INSERT, UPDATE, UPSERT, DELETE) made it as a flowfile attribute so that incoming flowfile can be decided which operation type to put in or use a single operation type.

SandishKumarHN commented:
@granthenke Added All Kudu Supported Predicates for KuduScan Processor

granthenke commented on nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-processors/src/main/java/org/apache/nifi/processors/kudu/ScanKudu.java:
```diff
@@ -0,0 +1,375 @@
+        try {
+            this.kuduTable = getKuduClient().openTable(tableName);
+        } catch (Exception e) {
+            e.printStackTrace();
```
Allow this exception to bubble up.

granthenke commented on nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-processors/src/main/java/org/apache/nifi/processors/kudu/AbstractKuduProcessor.java:
```diff
@@ -0,0 +1,451 @@
+            scannerBuilder.setProjectedColumnNames(projectedColumnNames);
+        }
+
+        AsyncKuduScanner scanner = scannerBuilder.build();
```
Can you upgrade to the Kudu 1.10.0 client and use a Java for each loop on the scanner? This will simplify scanning and handle scanner keep alive calls for you.

SandishKumarHN replied:
I did not understand this

granthenke replied:
In Kudu 1.10 instead of doing `while (scanner.hasMoreRows())`, you can do:

```
for (RowResult row : scanner) {
...
}
```

Though I think you need to use the standard KuduScanner, which I would recommend anyway. See above where I suggest building a standard KuduClient instead of the async one.

granthenke commented on nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-processors/src/main/java/org/apache/nifi/processors/kudu/AbstractKuduProcessor.java:
```diff
@@ -0,0 +1,451 @@
+        }
+    }
+
+    private void addPredicate(AsyncKuduScanner.AsyncKuduScannerBuilder scannerBuilder, KuduTable kuduTable, String column, String value, String comparisonOp) {
```
Kudu 1.10 has a generic addPredicate that can take an object. That might simplify things here.

granthenke commented on nifi-nar-bundles/nifi-kudu-bundle/nifi-kudu-processors/src/main/java/org/apache/nifi/processors/kudu/AbstractKuduProcessor.java:
```diff
@@ -0,0 +1,451 @@
+     * }
+     */
+    protected String convertToJson(Iterator<RowResult> rows) {
+        final StringBuilder jsonBuilder = new StringBuilder();
```
Is this the style other reading type processors use to build JSON? Is there a library that is more elegant than appending strings with StringBuilder?

granthenke replied:
Thinking about this a bit more, but not researching it. Would it be easier and more consistent to convert to a Nifi Record and use Nifi to convert to the needed formats?

SandishKumarHN replied:
@granthenke I have just followed the Hbase, Druid Processor's

SandishKumarHN commented:
@granthenke Thanks for the second review, WIll make changes as suggested!

SandishKumarHN commented:
@granthenke I have split this PR into two.
Kudu Put Operations: https://github.com/apache/nifi/pull/3610
Kudu Scana Processor:  https://github.com/apache/nifi/pull/3611

also made changes suggested here into new PR's

granthenke commented:
Thanks for breaking this out. I will review them asap.

SandishKumarHN commented:
@granthenke  made changes suggestion on the main PR. #3611

granthenke commented:
Can you close this PR now that it has been split out?

SandishKumarHN commented:
> Can you close this PR now that it has been split out?

Okay
